### PR TITLE
Allow rank computation for QRPivoted matrices 

### LINKED ISSUE #1052: rank(::QRPivoted) method?
Since pivoted QR is a rank-revealing calculation, it seems like we should have a `LinearAlgebra.rank` method that works directly on it.  Something like:
```jl
function rank(F::QRPivoted; atol::Real=0, rtol::Real=min(size(F)...)*eps(real(float(one(eltype(F)))))*iszero(atol))
    m = min(size(F)...)
    m == 0 && return 0
    tol = max(atol, rtol*abs(F.R[1,1]))
    return something(findfirst(i -> abs(F.R[i,i]) <= tol, 1:m), m+1) - 1
end
```
should do it, no?

Of course, this won't necessarily return the same thing as the SVD-based rank (and is less reliable than SVD — it shouldn't be the default method for `::AbstractMatrix`), but inconsistency is the fate of all numerical rank computations.  In typical cases it should be close, one hopes, and if you have the QR you should be able to explicitly ask what numerical rank it reveals.

### LINKED ISSUE #53214: rank(::QRPivoted) method?
Since pivoted QR is a rank-revealing calculation, it seems like we should have a `LinearAlgebra.rank` method that works directly on it.  Something like:
```jl
function rank(F::QRPivoted; atol::Real=0, rtol::Real=min(size(F)...)*eps(real(float(one(eltype(F)))))*iszero(atol))
    m = min(size(F)...)
    m == 0 && return 0
    tol = max(atol, rtol*abs(F.R[1,1]))
    return something(findfirst(i -> abs(F.R[i,i]) <= tol, 1:m), m+1) - 1
end
```
should do it, no?

Of course, this won't necessarily return the same thing as the SVD-based rank (and is less reliable than SVD — it shouldn't be the default method for `::AbstractMatrix`), but inconsistency is the fate of all numerical rank computations.  In typical cases it should be close, one hopes, and if you have the QR you should be able to explicitly ask what numerical rank it reveals.

### PR DESCRIPTION
**Pull request by danilonumeroso**
Hi, this fixes JuliaLang/LinearAlgebra.jl#1052. I followed @stevengj's code suggestion which worked just fine and added tests in the related section. 

Note: this is my first contribution so please feel free to point out any rookie mistakes I might have made!

### CODE REVIEW

**Comment by stevengj** [comment]
LGTM.  Do you also want to tackle https://github.com/JuliaLang/julia/issues/53214#issuecomment-1962035828?  If `nullspace` and `cond` aren't addressed in this PR, which is fine, we should open a new issue for those after this PR is merged.

**Review by stevengj — APPROVED** [review]

**Comment by stevengj** [comment]
Probably worth adding a `NEWS.md` entry.

**Comment by danilonumeroso** [comment]
> LGTM. Do you also want to tackle [#53214 (comment)](https://github.com/JuliaLang/julia/issues/53214#issuecomment-1962035828)? If `nullspace` and `cond` aren't addressed in this PR, which is fine, we should open a new issue for those after this PR is merged.

Yes, I'd like to. I'm going to open a new issue for those and tackle it in a separate PR

> Probably worth adding a `NEWS.md` entry.
> 
> Not sure where else it would be good to document this?

Done! Please tell me if I need to be more specific and/or there's any standard I should follow for NEWS.md entries

File: NEWS.md [hunk 1]
```diff
@@ -87,6 +87,8 @@ Standard library changes
 
 #### LinearAlgebra
 
+* `rank` can now take a QRPivoted matrix ([#53214]) to allow rank computation on QR-factorized matrices.
```
Comment (barucden) [inline NEWS.md:90]: The news item should cross reference the PR, not the issue.

Comment (danilonumeroso) [reply NEWS.md]: My bad, it should be ok now

Comment (stevengj) [inline NEWS.md]:
```suggestion
* `rank` can now take a `QRPivoted` matrix to allow rank estimation via QR factorization ([#53214]).
```

Comment (danilonumeroso) [reply NEWS.md]: Done, LMK whether I should merge after this

**Pull request merged**